File: Cpp/1338_reduce_array_size_to_the_half.cpp
// https://leetcode.cn/problems/reduce-array-size-to-the-half/

class Solution {
public:
    int minSetSize(vector<int>& arr) {
        unordered_map<int, int> cnts;
        for (int val: arr) {
            cnts[val] += 1;
        }
        priority_queue<int> pq;
        for (auto it = cnts.begin(); it != cnts.end(); ++it) {
            pq.push(it->second);
        }
        int res = 0, half = (arr.size() + 1) / 2, cnt = 0;
        while (cnt < half) {
            cnt += pq.top();
            pq.pop();
            res += 1;
        }
        return res;
    }
};

File: Python3/1338_reduce_array_size_to_the_half.py
# https://leetcode.cn/problems/reduce-array-size-to-the-half/

class Solution:
    def minSetSize(self, arr):
        """
        Inputs:
            arr: list[int]
        Outputs:
            res: int
        """
        cnts = defaultdict(int)
        for val in arr:
            cnts[val] += 1
        pq = []
        for val, cnt in cnts.items():
            pq.append(-cnt)
        heapq.heapify(pq)

        res = 0
        half, cnt = (len(arr)+1)//2, 0
        while cnt < half:
            cnt -= heapq.heappop(pq)
            res += 1
        return res